/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

syntax = "proto3";

package resdb.raft;

import "platform/proto/resdb.proto";
import "platform/proto/replica_info.proto";

// Replicated log entry used by RAFT.
message LogEntry {
  uint64 term = 1;
  uint64 index = 2;
  resdb.Request request = 3;
}

message AppendEntriesRequest {
  uint64 term = 1;
  uint32 leader_id = 2;
  uint64 prev_log_index = 3;
  uint64 prev_log_term = 4;
  repeated LogEntry entries = 5;
  uint64 leader_commit = 6;
}

message AppendEntriesResponse {
  uint64 term = 1;
  bool success = 2;
  uint64 match_index = 3;
  uint32 responder_id = 4;
  uint64 conflict_index = 5;
  uint64 conflict_term = 6;
}

message RequestVoteRequest {
  uint64 term = 1;
  uint32 candidate_id = 2;
  uint64 last_log_index = 3;
  uint64 last_log_term = 4;
}

message RequestVoteResponse {
  uint64 term = 1;
  bool vote_granted = 2;
  uint32 responder_id = 3;
}

message SnapshotMetadata {
  uint64 last_included_index = 1;
  uint64 last_included_term = 2;
  bytes state_hash = 3;
  resdb.ReplicaInfo leader_info = 4;
}

message SnapshotChunk {
  uint64 chunk_index = 1;
  bytes data = 2;
  bool last_chunk = 3;
}

message InstallSnapshotRequest {
  uint64 term = 1;
  uint32 leader_id = 2;
  SnapshotMetadata metadata = 3;
  SnapshotChunk chunk = 4;
}

message InstallSnapshotResponse {
  uint64 term = 1;
  bool success = 2;
  uint32 responder_id = 3;
  uint64 applied_index = 4;
}

message PersistentState {
  uint64 current_term = 1;
  uint32 voted_for = 2;
  uint64 commit_index = 3;
  uint64 last_applied = 4;
  SnapshotMetadata snapshot = 5;
}
